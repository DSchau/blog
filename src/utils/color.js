const getHashCode = str => {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  return hash;
};

const getHSL = (number, saturation, lightness) => {
  var shortened = Math.abs(number % 360);
  return `hsl(${shortened}, ${saturation}%, ${lightness}%)`;
};

export const getColorFromString = (str, lightness = 20, saturation = 90) => {
  return getHSL(getHashCode(str), saturation, lightness);
};
