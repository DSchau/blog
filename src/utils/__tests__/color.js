import * as color from '../color'

describe('getColorFromString', () => {
  const { getColorFromString } = color
  it('returns an HSL string from a string', () => {
    expect(getColorFromString('hello world')).toMatch(/hsl\(.+\)/)
  })

  it('can modify lightness', () => {
    expect(getColorFromString('hello world', 100)).toMatch('100%')
  })

  it('can modify saturation', () => {
    expect(getColorFromString('hello world', 0, 100)).toMatch('100%')
  })

  it('matches snapshot', () => {
    const strings = [
      `hello world`,
      `about`,
      `front end frameworks`,
      `getting started`,
      `hola`
    ].map(string => [string, getColorFromString(string)])

    expect(strings).toMatchSnapshot()
  })
})
