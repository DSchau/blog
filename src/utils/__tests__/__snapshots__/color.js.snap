// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getColorFromString matches snapshot 1`] = `
Array [
  Array [
    "hello world",
    "hsl(332, 90%, 20%)",
  ],
  Array [
    "about",
    "hsl(29, 90%, 20%)",
  ],
  Array [
    "front end frameworks",
    "hsl(231, 90%, 20%)",
  ],
  Array [
    "getting started",
    "hsl(227, 90%, 20%)",
  ],
  Array [
    "hola",
    "hsl(60, 90%, 20%)",
  ],
]
`;
